plugins/source/mixpanel/client/client.go

```diff
@@ -0,0 +1,116 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/cloudquery/cloudquery/plugins/source/mixpanel/internal/mixpanel"
+	"github.com/cloudquery/plugin-sdk/backend"
+	"github.com/cloudquery/plugin-sdk/plugins/source"
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/cloudquery/plugin-sdk/specs"
+	"github.com/rs/zerolog"
+)
+
+type Client struct {
+	logger     zerolog.Logger
+	sourceSpec specs.Source
+
+	MPSpec   Spec
+	Services *mixpanel.Client
+	Backend  backend.Backend
+}
+
+func New(logger zerolog.Logger, sourceSpec specs.Source, mpSpec Spec, services *mixpanel.Client, bk backend.Backend) Client {
+	return Client{
+		logger:     logger,
+		sourceSpec: sourceSpec,
+		MPSpec:     mpSpec,
+		Services:   services,
+		Backend:    bk,
+	}
+}
+
+func (c *Client) Logger() *zerolog.Logger {
+	return &c.logger
+}
+
+func (c *Client) ID() string {
+	return c.sourceSpec.Name
+}
+
+func Configure(ctx context.Context, logger zerolog.Logger, s specs.Source, o source.Options) (schema.ClientMeta, error) {
+	mpSpec := &Spec{}
+	if err := s.UnmarshalSpec(mpSpec); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal mixpanel spec: %w", err)
+	}
+
+	services, err := getServiceClient(logger, mpSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	cl := New(logger, s, *mpSpec, services, o.Backend)
+	return &cl, nil
+}
+
+func getServiceClient(logger zerolog.Logger, spec *Spec) (*mixpanel.Client, error) {
+	if spec.Secret == "" {
+		// allow empty username for the deprecated project secret method: https://developer.mixpanel.com/reference/project-secret
+		return nil, errors.New("no credentials provided")
+	}
+	if spec.ProjectID < 1 {
+		return nil, errors.New("no project id provided")
+	}
+
+	const dateFormat = "2006-01-02"
+
+	if spec.StartDate == "" {
+		dt := time.Now().UTC().Add(-30 * 86400 * time.Second).Format(dateFormat)
+		logger.Info().Str("start_date", dt).Msg("no start date provided, defaulting to 30 days ago")
+		spec.StartDate = dt
+	}
+	if spec.EndDate == "" {
+		dt := time.Now().UTC().Add(-86400 * time.Second).Format(dateFormat)
+		logger.Info().Str("end_date", dt).Msg("no end date provided, defaulting to yesterday")
+		spec.EndDate = dt
+	}
+
+	if _, err := time.Parse(dateFormat, spec.StartDate); err != nil {
+		return nil, fmt.Errorf("invalid start date format: %w", err)
+	}
+	if _, err := time.Parse(dateFormat, spec.EndDate); err != nil {
+		return nil, fmt.Errorf("invalid end date format: %w", err)
+	}
+
+	if spec.Timeout < 1 {
+		spec.Timeout = 30
+	}
+	if spec.MaxRetries < 1 {
+		spec.MaxRetries = 5
+	}
+
+	rg, err := mixpanel.ParseRegion(spec.Region)
+	if err != nil {
+		return nil, err
+	}
+	if rg == mixpanel.RegionNone {
+		rg = mixpanel.RegionUS
```

Comment:
What does region mean in the Mixpanel world? Does every workspace have a specific region? Could you miss datapoints if you extract from one region but actually multiple are in use?

Reply:
Your project lives in a specific region (EU residency data concerns). Multiple regions can't be in use in the same project. "documented" when creating a project :)

<img width="395" alt="Screenshot 2023-01-23 at 11 05 41" src="https://user-images.githubusercontent.com/223029/214024689-6e9bff1e-fd64-4fb0-be03-51eb7a23063b.png">